domain/domain.go

```diff
@@ -92,78 +92,75 @@ type Domain struct {
 	isLostConnectionToPD sync2.AtomicInt32 // !0: true, 0: false.
 }
 
-// loadInfoSchema loads infoschema at startTS into handle, usedSchemaVersion is the currently used
-// infoschema version, if it is the same as the schema version at startTS, we don't need to reload again.
-// It returns the latest schema version, the changed table IDs, whether it's a full load and an error.
-func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion int64,
-	startTS uint64) (neededSchemaVersion int64, change *tikv.RelatedSchemaChange, fullLoad bool, err error) {
+// loadInfoSchema loads infoschema at startTS.
+// It returns:
+// 1. the needed infoschema
+// 2. cache hit indicator
+// 3. currentSchemaVersion(before loading)
+// 4. the changed table IDs if it is not full load
+// 5. an error if any
+func (do *Domain) loadInfoSchema(startTS uint64) (infoschema.InfoSchema, bool, int64, *tikv.RelatedSchemaChange, error) {
 	snapshot := do.store.GetSnapshot(kv.NewVersion(startTS))
 	m := meta.NewSnapshotMeta(snapshot)
-	neededSchemaVersion, err = m.GetSchemaVersion()
+	neededSchemaVersion, err := m.GetSchemaVersion()
 	if err != nil {
-		return 0, nil, fullLoad, err
-	}
-	if usedSchemaVersion != 0 && usedSchemaVersion == neededSchemaVersion {
-		return neededSchemaVersion, nil, fullLoad, nil
+		return nil, false, 0, nil, err
 	}
 
-	// Update self schema version to etcd.
```

Comment:
Why remove this part?

Reply:
It is moved out of `loadInfoSchema()`. It is now in `Reload()`. I don't like this `defer` due to reassignment to `err`. And If we are going to improve the cache in the future, I guess it will probably become a low-level function that should not involve synchronization.